windows/src/global/inc/keymanengine.h

```diff
@@ -176,6 +196,29 @@ char* Debug_UnicodeString(PWSTR s, int x = 0);
 
 BOOL ShouldDebug_1(); // TSDMState state);
 
+// On failed condition it will always log "message" and return "retValue"
+// it will also assert if a debugger is attached for a debug build.
+#define DebugAssertRetValue(condition, message, retValue)               \
+  {                                                                     \
```

Comment:
Now I think about it, I'm not very keen on the macro having the return side-effect. It makes it harder for unfamiliar code readers to understand what is going on. Macros that impact control flow are probably the most difficult to work with.

Perhaps, instead, we have `DebugAssert` return TRUE/FALSE on the basis of `condition` and then the calling code is still in control of the failure case, e.g.:

```
  if(!DebugAssert(myCondition, message)) {
    return FALSE; // or some value
  }
```

Reply:
Also: I don't think we need `va_args` for debug assertions, so just the one macro should be fine.

Reply:
Ok I am a bit stumped on getting this working because macros really _replace_ with the code. So I can't really write the macro to work the way you have said.  I guess we could make a Globals::DebugAssert().  
Or when I did some googling there was this suggestion but that was what I already had https://stackoverflow.com/questions/7004327/is-it-bad-practice-to-have-a-macro-definition-return-a-value-for-a-function

Reply:
Yeah, you'd need an inline function, kinda like we do with `SendDebugMessage`. Use the macro to capture `__FILE__` etc and then rely on the inline function to do the work:

```
#define DebugAssert(cond,message) DebugAssert_1((cond),(message),__FILE__,__LINE__)

BOOL DebugAssert_1(BOOL cond, char *message, char *file, int line) {
 ...
}
```

Or something like that